- fix for memory corruption: on windows use _aligned_free() for memor…

guschmue opened the pull request:
…y allocated with aligned_alloc()
- enabled sparse_matmul and immutable_constant_op
- don't print LD_LIBRARY_PATH on error for windows (its null and crashes)
  with this python kernel tests pass (well, some error left related to stl float Inf)

Commit:
- fix for memory corruption: on windows use _aligned_free() for memory allocated with aligned_alloc()
- enabled sparse_matmul and immutable_constant_op
- don't print LD_LIBRARY_PATH on error for windows (its null and crashes)
with this python kernel tests pass (well, some error left related to stl float Inf)

tensorflow-jenkins commented:
Can one of the admins verify this patch?

drpngx commented:
Jenkins, test this please.

drpngx commented on tensorflow/contrib/cmake/tf_core_kernels.cmake:
```diff
@@ -85,33 +85,34 @@ list(REMOVE_ITEM tf_core_kernels_srcs ${tf_core_kernels_exclude_srcs})
+
+  # if we build for gpu include those  
+  list(APPEND tf_core_kernels_srcs 
+     "${tensorflow_source_dir}/tensorflow/core/kernels/debug_ops.h"
```
So I guess the windows behavior will be different from the linux one? One has debug ops, only in GPU, and the other one doesn't? We probably want the debug ops to be included (or not) not matter orthogonally to whether we're win32 or gpu.

drpngx commented on tensorflow/core/common_runtime/gpu/pool_allocator.h:
```diff
@@ -173,7 +173,11 @@ class BasicCPUAllocator : public SubAllocator {
   void* Alloc(size_t alignment, size_t num_bytes) override {
     return port::aligned_malloc(num_bytes, alignment);
   }
+#if defined(PLATFORM_WINDOWS)
```
there should be a port::aligned_free which is valid for all cases. ASAN or valgrind should be able to flag that.

guschmue replied:
yes, there is  - let me switch to this.

drpngx commented on tensorflow/core/kernels/debug_ops.h:
```diff
@@ -127,7 +130,16 @@ class DebugNanCountOp : public OpKernel {
+        // with integral values, so we explicitly convert an integral value to
+        // double before calling isnan.
+        typedef std::conditional<std::is_integral<T>::value, double, T>::type R;
+        R value = static_cast<R>(input_flat[i]);
```
```const R value```

Would just `static_cast<double>(input_flat[i])` work with all types?

guschmue replied:
let me look for  a better way

drpngx commented on tensorflow/core/kernels/debug_ops.h:
```diff
@@ -127,7 +130,16 @@ class DebugNanCountOp : public OpKernel {
+        typedef std::conditional<std::is_integral<T>::value, double, T>::type R;
+        R value = static_cast<R>(input_flat[i]);
+#else
+        T& value = input_flat[i];
```
const

guschmue replied:
ok

drpngx commented on tensorflow/core/kernels/sparse_matmul_op.h:
```diff
@@ -18,6 +18,22 @@ limitations under the License.
 
 #include "third_party/eigen3/Eigen/Core"
 #include "tensorflow/core/platform/types.h"
+#ifdef _MSC_VER
```
is it possible to find a better home for this, somewhere in platform? Maybe platform/windows/intrinsics_port.h?

guschmue replied:
fair.

drpngx reviewed (changes requested):
Nice!

drpngx commented:
Woops, it doesn't like `std::bind` and other things. Please take a look at the errors:

https://ci.tensorflow.org/job/tensorflow-pull-requests-multijob/2416/

Also, if you're comfortable with git, you could split the commits.

guschmue commented:
ah, should have tried on linux first. Working on it.

guschmue commented on tensorflow/contrib/cmake/tf_core_kernels.cmake:
```diff
@@ -85,33 +85,34 @@ list(REMOVE_ITEM tf_core_kernels_srcs ${tf_core_kernels_exclude_srcs})
-      "${tensorflow_source_dir}/tensorflow/core/kernels/sparse_matmul_op.h"
       "${tensorflow_source_dir}/tensorflow/core/kernels/*quantiz*.h"
       "${tensorflow_source_dir}/tensorflow/core/kernels/*quantiz*.cc"
+      "${tensorflow_source_dir}/tensorflow/core/kernels/meta_support.*"
```
there was a mege conflict on meta_support

guschmue commented on tensorflow/contrib/cmake/tf_core_kernels.cmake:
```diff
@@ -85,33 +85,34 @@ list(REMOVE_ITEM tf_core_kernels_srcs ${tf_core_kernels_exclude_srcs})
-   "${tensorflow_source_dir}/tensorflow/core/kernels/*.cu.cc"
-   "${tensorflow_source_dir}/tensorflow/contrib/rnn/kernels/*.cu.cc"
-)
+if(WIN32 AND tensorflow_ENABLE_GPU)
```
it was collecting the files in tf_core_gpu_kernels_srcs but for cpu no target use this variable. That's maybe ok but its risky breaking somebody else using cmake so I figured its better to ifdef it.

guschmue commented on tensorflow/contrib/cmake/tf_core_kernels.cmake:
```diff
@@ -85,33 +85,34 @@ list(REMOVE_ITEM tf_core_kernels_srcs ${tf_core_kernels_exclude_srcs})
   list(REMOVE_ITEM tf_core_gpu_kernels_srcs ${tf_core_gpu_kernels_exclude_srcs})
-endif(WIN32)
+
+  # if we build for gpu include those  
```
those don't compile yet, some ambiguous operator error when compiling with msvc. I have a fix but I'm looking for a better fix.

guschmue commented on tensorflow/core/platform/test.cc:
```diff
@@ -13,6 +13,10 @@ See the License for the specific language governing permissions and
 ==============================================================================*/
 
+#ifdef PLATFORM_WINDOWS
+#include <Windows.h>
```
ok

guschmue commented:
looking over this, you are right: let me split this into a couple of commits. That also helps to get this aligned_free fix in.

guschmue commented:
closing this one and file some more targeted pr's instead. Going to reference this pr so we don't lose the review feedback.

drpngx commented:
Please merge with #5207, which will probably create some conflicts on the `meta_support.*`

guschmue commented:
I closed this pr and split it - too many things in it. I think I'll split it in 4 prs:
1. the memory corruption (addressed your review comments and send a new pr  #5288)
2. enable sparse mat ops (nearly done)
3. enable immutable ops
4. enable debug_op 
I'll make sure sure they are on top of the meta_support.* changes 
Hope that works.

drpngx commented:
nice, thanks!

drpngx commented:
BTW, for the `std::bind` thing, we're seeing an error in #5286, possibly the easiest fix would be to use `SparseSlice<TL>*` instead of `auto*`.

drpngx commented:
Oh, BTW, flipping through some code today, it occurred to me that we should probably `#define WIN32_LEAN_AND_MEAN 1` before including windows.h headers. I usually don't like to have them in header files, only in cpp files, because there a ton of symbols in there.

guschmue commented:
Ah, lean and mean we have: gets added in CMakeLists.txt to the compiler defines for windows.
Going to try SparseSlice<TL>* shortly. Would love to get rid of the std::bind() without a ifdef.

guschmue commented:
turns out I did not need the std::bind(). We had made this change before nvidia support vs2015.update3 (we used vs2015.update1 until the official cuda 8.0 came out). Now that the std:bind() made trouble for gcc I started over on sparse_maxmul_op.cc and its fine without std:bind(): builds on linux and windows.

drpngx commented:
Ah, OK, great to hear! Thanks for making these changes!
